content/blog/2019-11-testthat-2-3-0/index.Rmarkdown

```diff
@@ -0,0 +1,142 @@
+
+```{r, eval = FALSE}
+f <- function() g()
+g <- function() h()
+h <- function() stop("This is an error!")
+
+test_that("f() works as expected", {
+  expect_equal(f(), 10)
+})
+```
+
+You'll now get an informative backtrace that should allow you to quickly locate the source of the error:
+
+```
+test-error.R:6: error: f() works as expected
+This is an error!
+Backtrace:
+ 1. testthat::expect_equal(f(), 10) tests/testthat/test-error.R:6:2
+ 4. testthat:::f()
+ 5. testthat:::g() tests/testthat/test-catch.R:1:5
+ 6. testthat:::h() tests/testthat/test-catch.R:2:5
+```
+
+The previous version only showed the error message, which wasn't terribly useful!
+
+## `verify_output()`
+
+`verify_output()` is designed to test output aimed at a human, like print methods and error messages.
+Here you want to test that the output is useful to a human, but there's obviously no way to do that automatically.
+Instead, the best you can do is to check the results with your eyeballs every time the results change.
+`verify_output()` is designed to help you do this, making it a type of visual regression test.
+
+You'll need to use `verify_output()` in concert with git.
+Whenever the output changes, you'll get a test failure, but to see the change, you'll need to use git.
+If the change is correct, commit it with git; if it's incorrect, revert it.
+
+`verify_output()` works a little like RMarkdown: you give it some R code and it will run it and interleave the input and ouptut.
+For example, imagine we were writing some tests to check that tibbles print correctly:
+
+```{r, include = FALSE}
+test_path <- function(x) x
+unlink("test-print-dataframe.txt")
+unlink("test-print-dataframe-2.txt")
+```
+
+```{r}
+library(tibble)
+
+test_that("tibbles print usefully", {
+  verify_output(test_path("test-print-dataframe.txt"), {
+    df1 <- tibble(x = 1:1e6)
+    print(df1)
+  })
+})
+```
+
+That test yields a `test-print-dataframe.txt` containing this output:
+
+```{r, results = "asis", echo = FALSE}
+cat("```\n")
+writeLines(readLines("test-print-dataframe.txt"))
+cat("```\n")
+```
+
+Unfortunately there's no way for `verify_output()` to capture comments, so you can instead use bare strings if you want comments to appear in the output. If you start the comment with `#` it will be formatted as a heading:
+
+```{r}
+test_that("tibbles print usefully", {
+  verify_output(test_path("test-print-dataframe-2.txt"), {
+    "# long tibbles"
+    df1 <- tibble(x = 1:1e6)
+    print(df1)
+    
+    "# wide tibbles"
+    "not yet written"
+  })
+})
+```
+
+```{r, results = "asis", echo = FALSE}
+cat("```\n")
+writeLines(readLines("test-print-dataframe-2.txt"))
+cat("```\n")
+```
+
+`verify_output()` is automatically skipped when run on CRAN.
+This avoids false positives because it's very easy to accidentally depend on the code from other packages, and failure does not imply incorrect computation, just a change in presentation.
```

Comment:
I get the "failure does not imply incorrect computation, just a change in presentation" portion above, but not the tie in to depending on code from other packages.

Reply:
@lionel-'s explanation makes sense to me, though. So, resolved.